File: packages/core/cpp/nativecompose/threadedruntime/ThreadedRuntimeBindingsInstaller.cpp
#include "ThreadedRuntimeBindingsInstaller.h"

#ifdef __ANDROID__

namespace nativecompose::threadedruntime {

namespace jsi = facebook::jsi;

facebook::jni::local_ref<JThreadedRuntimeBindingsInstaller::jhybriddata>
JThreadedRuntimeBindingsInstaller::initHybrid(
    facebook::jni::alias_ref<jclass>,
    std::string runtimeName,
    std::string kind) {
  return makeCxxInstance(std::move(runtimeName), std::move(kind));
}

void JThreadedRuntimeBindingsInstaller::registerNatives() {
  registerHybrid({
      makeNativeMethod(
          "initHybrid", JThreadedRuntimeBindingsInstaller::initHybrid),
  });
}

facebook::react::BindingsInstaller::BindingsInstallFunc
JThreadedRuntimeBindingsInstaller::getBindingsInstallFunc() {
  return [runtimeName = runtimeName_, kind = kind_](jsi::Runtime& runtime) {
    auto env = jsi::Object(runtime);
    env.setProperty(
        runtime, "kind", jsi::String::createFromUtf8(runtime, kind));
    env.setProperty(
        runtime,
        "runtimeName",
        jsi::String::createFromUtf8(runtime, runtimeName));
    runtime.global().setProperty(runtime, "__THREADED_RUNTIME_ENV__", env);
  };
}

} // namespace nativecompose::threadedruntime

#endif // __ANDROID__

File: packages/core/android/src/main/java/com/nativecompose/threadedruntime/ThreadedRuntimeBindingsInstaller.kt
package com.nativecompose.threadedruntime

import com.facebook.jni.HybridData
import com.facebook.react.runtime.BindingsInstaller

/**
 * Installs `__THREADED_RUNTIME_ENV__` into a secondary runtime's jsi::Runtime
 * during ReactInstance initialization — before any script (including Metro
 * polyfills) evaluates, and regardless of whether the bundle comes from the
 * delegate loader (release) or dev support (debug + Metro). Android
 * counterpart of the JSI injection iOS does in `host:didInitializeRuntime:`.
 */
internal class ThreadedRuntimeBindingsInstaller(runtimeName: String, kind: String) :
    BindingsInstaller(initHybrid(runtimeName, kind)) {
  private companion object {
    init {
      com.margelo.nitro.threadedruntime.NativeComposeThreadedRuntimeOnLoad.initializeNative()
    }

    @JvmStatic private external fun initHybrid(runtimeName: String, kind: String): HybridData
  }
}
